runtime/compiler/runtime/JITServerAOTDeserializer.cpp

```diff
@@ -898,3 +901,593 @@ JITServerLocalSCCAOTDeserializer::updateSCCOffsets(SerializedAOTMethod *method,
+   }
+
+void
+JITServerNoSCCAOTDeserializer::invalidateMethod(J9Method *method)
+   {
+   auto p_it = _methodPtrMap.find(method);
+   if (p_it == _methodPtrMap.end()) // Not cached or already marked as invalid
+      return;
+
+   uintptr_t id = p_it->second;
+   auto i_it = _methodIdMap.find(id);
+   TR_ASSERT(i_it != _methodIdMap.end(), "Broken two-way map");
+
+   i_it->second = NULL;
+
+   _methodPtrMap.erase(p_it);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Invalidated RAMMethod %p ID %zu", method, id);
+   }
+
+void
+JITServerNoSCCAOTDeserializer::clearCachedData()
+   {
+   _classLoaderIdMap.clear();
+   _classLoaderPtrMap.clear();
+
+   _classIdMap.clear();
+   _classPtrMap.clear();
+
+   _methodIdMap.clear();
+   _methodPtrMap.clear();
+
+   for (auto &kv : _classChainMap)
+      TR::Compiler->persistentGlobalMemory()->freePersistentMemory(kv.second);
+   _classChainMap.clear();
+
+   for (auto &kv : _wellKnownClassesMap)
+      TR::Compiler->persistentGlobalMemory()->freePersistentMemory(kv.second);
+   _wellKnownClassesMap.clear();
+
+   getNewKnownIds().clear();
+   }
+
+bool
+JITServerNoSCCAOTDeserializer::cacheRecord(const ClassLoaderSerializationRecord *record, TR::Compilation *comp, bool &isNew, bool &wasReset)
+   {
+   OMR::CriticalSection cs(getClassLoaderMonitor());
+   if (deserializerWasReset(comp, wasReset))
+      return false;
+
+   auto it = _classLoaderIdMap.find(record->id());
+   if (it != _classLoaderIdMap.end())
+      return true;
+   isNew = true;
+
+   // Lookup the class loader using the name of the first class that it loaded
+   auto loader = (J9ClassLoader *)getLoaderTable()->lookupClassLoaderAssociatedWithClassName(record->name(), record->nameLength());
+   if (!loader)
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+            "ERROR: Failed to find class loader for first loaded class %.*s", RECORD_NAME(record)
+         );
+      return false;
+      }
+
+   addToMaps(_classLoaderIdMap, _classLoaderPtrMap, it, record->id(), loader, loader);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+         "Cached class loader record ID %zu -> { %p } for first loaded class %.*s",
+         record->id(), loader, RECORD_NAME(record)
+      );
+   return true;
+   }
+
+bool
+JITServerNoSCCAOTDeserializer::cacheRecord(const ClassSerializationRecord *record, TR::Compilation *comp, bool &isNew, bool &wasReset)
+   {
+   OMR::CriticalSection cs(getClassMonitor());
+   if (deserializerWasReset(comp, wasReset))
+      return false;
+
+   auto it = _classIdMap.find(record->id());
+   if (it != _classIdMap.end())
+      {
+      if (it->second._ramClass)
+         {
+         return true;
+         }
+      else
+         {
```

Comment:
Not at the moment. If class re-caching is added in future then these cases will need to be distinguished. Otherwise the entry simply needs to sit there in either case so the deserializer knows that that particular record ID is no longer valid.